[tools/depends] meson cross file generation during configure

fuzzard opened the pull request:
## Description
generate target meson cross file during configure. This generates at the same time as
toolchain and site config files.
Updated configure.ac to remove some bashisms. This allows configure script to be run on PURE posix based shells (eg DASH)

## Motivation and context
With recent changes to the generated Toolchain file, we do a find_file call looking for the meson cross file, and we have this set as required.
There is a case where a dependency that uses meson, and therefore generates the meson cross file isnt run before a dependency that uses the cmake toolchain file, and we then get a build failure for the cmake based dependency.

Ive no doubt @lrusak is going to cringe about this, but i cant come up with a cleaner way to handle this.
I had looked at generating as part of native dep's, but then we still want to store in target dep folder, and if we need to regenerate the cross it then requires going into native dep to trigger. It was messy.
Another option was to add a cmake-toolchain make target, with a dependency to meson-cross-file. This then added a whole bunch of defined targets with just meson-cross-file to force its generation before a cmake based project was executed. Again, this was messy because it added dependencies to something that wasnt relevant to that dependency.

for both of these situations, it also made it difficult to go in and trigger something like crossguid (ie cd tools/depends/target/crossguid && make) without already running make meson-cross-file. Not a normal thing for most, but i do these sort of things quite regularly when updating dependencies. more of a QOL aspect for myself.

So, we generate toolchain.cmake and also config.site as part of configure and populate target shared folder. to me it feels more natural to also generate meson cross file at the same point. It removes a dependency on python used in the existing script, and will stay up to date with toolchain and config.site being regenerated via configure.

If anyone can come up with another method, im all ears.

## How has this been tested?
Compare generation to previous python based implementation. Same output is generated for relevant flags/compilers

## What is the effect on users?
N/A

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

eli-schwartz commented on tools/depends/configure.ac:
```diff
@@ -718,6 +718,156 @@ AC_SUBST(app_rendersystem)
+[
+# Parse and handle CC
+read -r -a CC_ARRAY <<< "$CCACHE $CC"
+if [ ${#CC_ARRAY[@]} -eq 1 ]
```
You cannot use bashisms in a POSIX sh script (which is what configure.ac generates).

And here you are simultaneously using:
- bash arrays
- POSIX sh `[` (rather than bash-specific `[[`)
- bash herestring `<<<`
- pre-POSIX `[ "x$foo" = "xbar" ]` tricks for not bombing out when $foo evaluates to something that begins with a dash, and the `[` program is badly written and tries to check a 3-value `strval operator strval` usage using the two-value `-option strval` form

fuzzard replied:
feel free to give any info to make it more "correct". i only did enough to get it to function. worked on macos, worked on our linux CI hosts, so thats all i can test really

eli-schwartz replied:
It will work if your system happens to symlink `/bin/sh` to `bash`, which is not at all guaranteed.
It's definitely the case on macOS, but different linuxes do this differently, and some even make it a configuration choice.

Anyway, this PR moves some logic from both target/Makefile and from target/meson-cross-setup.sh, but maybe you could simply run the latter from configure.ac instead of inlining it? The standalone script appears to work fine as it calls out to python in order to do fiddly string splitting and array-izing, and therefore does not care what shell it runs in.

You could also document that anyone who gets really strange syntax errors trying to run `./configure` should mess with root-owned files in `/bin`, I suppose. :)

fuzzard replied:
the issue with the script, is it requires python. in a tools depend build, python is build as part of native tools, ie after configure, chicken and egg issue.

Can you provide me with details of how to repro a failure. Im ok to try and make things work, but i need to be able to repro such a setup so i can test and see how it fails.

eli-schwartz replied:
What's your development system? On Debian/Ubuntu, you could install the `dash` package. macOS Catalina has it preinstalled, apparently (and people have theorized as a result that Apple may someday stop using bash for /bin/sh by switching it to dash). In either case, you could run `dash configure` to manually override the interpreter you run the configure script with.

fuzzard replied:
yeah i run macos, but can VM whatever i need. I'll try the dash thing and see how it goes i guess.

fuzzard replied:
well dash doesnt even work with current master on macos, so im not going to bother fixing something that doesnt work anyway. i'll try an ubuntu vm.

Can i also just confirm, tools/depends really is only intended to build apple or android platforms, not linux. what triggered you looking at this?

fuzzard replied:
ok, how about this @eli-schwartz 

https://github.com/fuzzard/xbmc/commit/9d3960a638c379331c63946f59c1c24620b43911

Does that work for your situation?

Commit:
[tools/depends] configure remove (some) "bashisms"

Commit:
[tools/depends] meson cross file generation during configure

generate target meson cross file during configure. This generates at the same time as
toolchain and site config files.

fuzzard commented:
@wsnipex not sure if this is in your wheel house, but thought id ping you for one last look. I'll merge this in a day or two otherwise, as we are getting build failures because of what this intends to fix. eg https://jenkins.kodi.tv/job/LINUX/40702/console

```
21:30:27 CMake Error at /home/jenkins/workspace/LINUX-64-GLES/tools/depends/xbmc-depends/x86_64-linux-gnu-debug/share/Toolchain.cmake:161 (find_file):
21:30:27   Could not find MESON-CROSS using the following files: cross-file.meson
21:30:27 Call Stack (most recent call first):
21:30:27   /home/jenkins/workspace/LINUX-64-GLES/tools/depends/xbmc-depends/x86_64-linux-gnu-native/share/cmake-3.21/Modules/CMakeDetermineSystem.cmake:124 (include)
21:30:27   CMakeLists.txt:4 (project)
21:30:27 
21:30:27 
21:30:27 CMake Error: CMake was unable to find a build program corresponding to "Unix Makefiles".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.
21:30:27 -- Configuring incomplete, errors occurred!
21:30:27 make[2]: *** [Makefile:19: x86_64-linux-gnu-debug] Error 1
```

wsnipex reviewed (approved):
looks ok to me. Makes sense to generate toolchains/configs for all the different build systems at the same time
